src/core/lib/promise/activity.h

```diff
@@ -0,0 +1,437 @@
+
+  // Return the current activity.
+  // Additionally:
+  // - assert that there is a current activity (and catch bugs if there's not)
+  // - indicate to thread safety analysis that the current activity is indeed
+  //   locked
+  // - back up that assertation with a runtime check in debug builds (it's
+  //   prohibitively expensive in non-debug builds)
+  static Activity* current() ABSL_ASSERT_EXCLUSIVE_LOCK(current()->mu_) {
+#ifndef NDEBUG
+    GPR_ASSERT(g_current_activity_);
+    if (g_current_activity_ != nullptr) {
+      g_current_activity_->mu_.AssertHeld();
+    }
+#endif
+    return g_current_activity_;
+  }
+
+  // Produce an activity-owning Waker. The produced waker will keep the activity
+  // alive until it's awoken or dropped.
+  Waker MakeOwningWaker() {
+    Ref();
+    return Waker(this);
+  }
+
+  // Produce a non-owning Waker. The waker will own a small heap allocated weak
+  // pointer to this activity. This is more suitable for wakeups that may not be
+  // delivered until long after the activity should be destroyed.
+  Waker MakeNonOwningWaker() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
+
+ protected:
+  inline ~Activity() override {
+    if (handle_) {
+      DropHandle();
+    }
+  }
+
+  // All promise execution occurs under this mutex.
+  Mutex mu_;
+
+  // Check if this activity is the current activity executing on the current
+  // thread.
+  bool is_current() const { return this == g_current_activity_; }
+  // Check if there is an activity executing on the current thread.
+  static bool have_current() { return g_current_activity_ != nullptr; }
+  // Check if we got an internal wakeup since the last time this function was
+  // called.
+  bool got_wakeup() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_) {
+    return absl::exchange(got_wakeup_during_run_, false);
+  }
+
+  // Set the current activity at construction, clean it up at destruction.
+  class ScopedActivity {
+   public:
+    explicit ScopedActivity(Activity* activity) {
+      GPR_ASSERT(g_current_activity_ == nullptr);
+      g_current_activity_ = activity;
+    }
+    ~ScopedActivity() { g_current_activity_ = nullptr; }
+    ScopedActivity(const ScopedActivity&) = delete;
+    ScopedActivity& operator=(const ScopedActivity&) = delete;
+  };
+
+  // Implementors of Wakeable::Wakeup should call this after the wakeup has
+  // completed.
+  void WakeupComplete() { Unref(); }
+
+ private:
+  class Handle;
+
+  void Ref() { refs_.fetch_add(1, std::memory_order_relaxed); }
+  void Unref() {
+    if (1 == refs_.fetch_sub(1, std::memory_order_acq_rel)) {
+      delete this;
+    }
+  }
+
+  // Return a Handle instance with a ref so that it can be stored waiting for
+  // some wakeup.
+  Handle* RefHandle() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
+  // If our refcount is non-zero, ref and return true.
+  // Otherwise, return false.
+  bool RefIfNonzero();
+  // Drop the (proved existing) wait handle.
+  void DropHandle() EXCLUSIVE_LOCKS_REQUIRED(mu_);
```

Comment:
prefer `ABSL_EXCLUSIVE_LOCKS_REQUIRED`. I won't comment on other missing `ABSL_` prefixes herein.

Reply:
If you spot them lmk... I've done a few passes through this and obviously missed some